torch/nn/utils/attention.py

```diff
@@ -0,0 +1,253 @@
+""" Define Base class as well as some crowd favorites """
+from abc import ABC, abstractmethod
+from enum import IntEnum
+from typing import Optional, Union
+from warnings import warn
+
+import torch
+from torch.backends.cuda import (
+    can_use_efficient_attention,
+    can_use_flash_attention,
+    SDPAParams,
+)
+from torch.nn.functional import scaled_dot_product_attention
+from torch.utils import _pytree as pytree
+
+
+__all__ = ["AttnBias", "TensorBias", "CausalVariant", "CausalBias"]
+
+
+def _input_requires_grad(*tensors: torch.Tensor) -> bool:
+    """Returns True if any of the tensors requires grad"""
+    return any(t.requires_grad for t in tensors)
+
+
+def _materialize_if_needed(
+    bias: "AttnBias", device: Optional[torch.device] = None
+) -> Union[torch.Tensor, "AttnBias"]:
+    if bias.needs_materialization():
+        return bias.materialize(device)
+    return bias
+
+
+def _postprocess_flash_output(inpt_tensor: torch.Tensor, og_size: int):
+    """Handles the unpad of the last dimension"""
+    if inpt_tensor.size(-1) != og_size:
+        return inpt_tensor[..., :og_size]
+    return inpt_tensor
+
+
+class AttnBias(ABC):
+    """Abstract base class for attention biases"""
+
+    @abstractmethod
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
+        raise NotImplementedError("This is an abstract base class")
+
+    @abstractmethod
+    def needs_materialization(self) -> bool:
+        raise NotImplementedError("This is an abstract base class")
+
+    @staticmethod
+    @abstractmethod
+    def dispatch(
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_mask,
+        dropout_p: float,
+        is_causal: bool,
+        scale: Optional[float],
+    ) -> torch.Tensor:
+        raise NotImplementedError("This is an abstract base class")
+
+
+class TensorBias(AttnBias):
+    """A bias that is a tensor"""
+
+    def __init__(self, bias: torch.Tensor):
+        self.bias = bias
+
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
+        return self.bias
+
+    def needs_materialization(self) -> bool:
+        return True
+
+    @staticmethod
+    def dispatch(
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_mask: "AttnBias",
+        dropout_p: float,
+        is_causal: bool,
+        scale: Optional[float],
+    ) -> torch.Tensor:
+        raise NotImplementedError(
+            "TensorBias requires materialization, so this should never be called!"
```

Comment:
so is the idea that materialization and dispatching are mutually exclusive in general?

do I have it right that materialization necessarily returns a standard dense Tensor so we aren't expected to ever dispatch on some custom type?

Reply:
I wouldn't say mutually exclusive but they are independent. the subclass defines how it should produce the sdpa_output using the `dispatch` method and it will use the base_class __torch_function__. 

There is a shortcut path for `needs_materialization` == True subclasses. I could make these classes implement dispatch. And in the dispatch make users define call `marterialize`. The idea was that users don't have to implement dispatch and just Raise the error.